Replace bitlab dependency with homegrown bit twiddling

djc opened the pull request:
bitlab was a bit of an odd dependency, providing some convenience wrappers for writing bits across byte slices. It additionally pulled in the num crate. Replace it by writing our own version of the bit reading/writing routines, which should amount to much less code and is more optimized for our use case since a number of checks can be elided.

I'm not a very experienced bit twiddler and it took me way too long to get the encode side working. Happily we have pretty good regression tests for this stuff! Any feedback is welcome.

Ralith commented on quinn-h3/src/qpack/prefix_string/decode.rs:
```diff
@@ -78,6 +75,23 @@ impl HuffmanDecoder {
+    byte_offset += bit_offset / 8;
+    bit_offset -= (bit_offset / 8) * 8;
+    Ok(if bit_offset + len <= 8 {
+        (u8::from_be(src[byte_offset as usize]) << bit_offset) >> (8 - len)
```
`u8::from_be` is the identity function; endianness only exists between bytes

djc replied:
Yeah, forgot to cut that out.

Ralith commented on quinn-h3/src/qpack/prefix_string/decode.rs:
```diff
@@ -78,6 +75,23 @@ impl HuffmanDecoder {
     }
 }
 
+fn extract_bits(src: &[u8], mut byte_offset: u32, mut bit_offset: u32, len: u32) -> Result<u8, ()> {
```
As someone who's also not a bit twiddling expert, I think this needs a decent battery of unit tests regardless of my opinion of the implementation.

djc replied:
Both `extract_bits()` and `set_bits()` are [fully covered](https://codecov.io/gh/djc/quinn/commit/4f5f85daeac3d8f12825e9ea31ea5a612897584f) by existing tests. As such, I don't see a whole lot of value in adding more. What do you think?

Ralith replied:
That link doesn't work here (and I can't seem to find equivalent information by browsing manually) but I think it's somewhat beside the point. It's useful to have narrowly scoped unit tests even if they're subsumed by larger ones, because it's easier to be sure that the narrowly scoped tests cover all interesting cases (which can be more subtle than just ensuring that every branch is traversed), and if something *is* wrong, the narrow tests are much more helpful for diagnosing an issue, both now and in the future (I'm reluctant to bet that we'll never want to touch a given piece of code again).

djc replied:
Done.

stammw commented:
Got this tested with QPACK interop test. Everything is fine :+1:

Ralith commented on quinn-h3/src/qpack/prefix_string/encode.rs:
```diff
@@ -93,6 +75,32 @@ impl HuffmanEncoder {
+        let pad_left = out[pos.byte as usize] | PAD_RIGHT[(8 - pos.bit) as usize];
+        let shifted = value << (8 - pos.bit - pos.count) | PAD_LEFT[pos.bit as usize];
+        let pad_right = PAD_RIGHT[(8 - pos.count - pos.bit) as usize];
+        out[pos.byte as usize] = (pad_left & shifted) | pad_right;
```
Is it intended that if bit = 0 and count = 8, `value` is completely ignored? I'm probably missing something here.

djc replied:
Pretty sure it isn't? I think the first part of the test covers this case.

djc replied:
As in this case:

```
pos.count = 8;
set_bits(&mut buf, &pos, 0b1_0101);
assert_eq!(&buf[..1], &[0b1_0101]);
pos.byte += 1;
```

Ralith replied:
Er, sorry, I meant when bit = 0 and count = 0.

Ralith replied:
For example:

```rust
let mut buf = [0; 1];
set_bits(
    &mut buf,
    &BitWindow {
        byte: 0,
        bit: 0,
        count: 1,
    },
    0,
);
```

produces `buf == [255]`, which doesn't seem right, unless I'm wildly misunderstanding the intended semantics (which should probably be commented?)

Ralith commented:
I'm prepared to believe that these functions are working as intended, but exactly what the intended semantics are is mysterious to me, particularly for `set_bits`. Could we have some comments describing what assumptions they make and why?

djc commented:
Yeah, I was thinking the same thing. I've renamed these functions to `read_bits()` and `write_bits()` and added a whole bunch of comments. Let me know if that makes things more clear!

Ralith commented on quinn-h3/src/qpack/prefix_string/decode.rs:
```diff
@@ -78,6 +75,29 @@ impl HuffmanDecoder {
 
+// Read `len` bits from the `src` slice at the specified position
+//
+// Never read more than 8 bits at a time. `bit_offset` may be larger than 8.
```
nit: these might as well be doc comments

djc replied:
Fixed.

Ralith commented on quinn-h3/src/qpack/prefix_string/encode.rs:
```diff
@@ -93,6 +75,43 @@ impl HuffmanEncoder {
+// always called prior to calling this function.
+//
+// The bits to be written to are expected to be set to 1 when calling this function. Similarly,
+// this function maintains the invariant that unused bits in the output bytes are set to 1.
```
Using 1 for unwritten data rather than 0 seems pretty unconventional. Is there a reason?

djc replied:
I think this is described [here](https://tools.ietf.org/html/rfc7541#appendix-B). Added a reference to this in the doc comment for `write_bits()`.

Ralith commented on quinn-h3/src/qpack/prefix_string/encode.rs:
```diff
@@ -93,6 +75,43 @@ impl HuffmanEncoder {
+// always called prior to calling this function.
+//
+// The bits to be written to are expected to be set to 1 when calling this function. Similarly,
+// this function maintains the invariant that unused bits in the output bytes are set to 1.
```
nit: doc comment

djc replied:
Fixed. (I thought clippy didn't like doc comments for private functions, but that doesn't appear to be an issue.)

Ralith replied:
In fact, rustdoc fully supports documenting such things via `--document-private-items`. Handy for exploring a codebase.
